pkg/ipcache/ipcache.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipcache
+
+import (
+	"encoding/json"
+	"path"
+	"sort"
+	"sync"
+
+	"github.com/cilium/cilium/pkg/envoy"
+	envoyAPI "github.com/cilium/cilium/pkg/envoy/cilium"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// DefaultAddressSpace is the address space used if none is provided.
+	// TODO - once pkg/node adds this to clusterConfiguration, remove.
+	DefaultAddressSpace = "default"
+)
+
+var (
+	// IPIdentitiesPath is the path to where endpoint IPs are stored in the key-value
+	//store.
+	IPIdentitiesPath = path.Join(kvstore.BaseKeyPrefix, "state", "endpointIPs", "v1")
```

Comment:
I have been thinking about this for  a long while. Let's rename this to just `ip` or `ipcache` as we might do IP to identity mapping for things other than what we call endpoints. Examples could be external IPs, or VM IPs. The purpose of this cache is to map IP to identity.

Due to the design, the cache could actually do LPM lookups and we could resolve subnets as well with some minimal added logic.